Allow PSyIR blocks to be generated inside f2pygen (closes #950)

sergisiso opened the pull request:
Provides a temporary mechanism to start porting the PSy-layers to PSyIR without the need to port it all at once.

It already allows to do things like module-inlining modified kenrels into the PSy-layer.

Closes #950

I think #490 and #663 may also be fixed by this.

sergisiso commented:
@rupertford and @arporter This still needs to be clean up and tested, but before going further, would in your opinion this new f2pygen for PSyIR blocks be a good approach while we don't have the full PSy-layer translated?

rupertford commented:
It looks like a good idea to me and already has benefit (in that modified PSyIR can be inlined). I'm afraid I can't really comment with any real certainty on how useful it might be as we migrate to using the PSyIR and the back ends without thinking about it in more depth, but it seems like it would help!

arporter commented:
I like it too. It's only a small change and it's already solved a significant limitation (the fact that module in-lining can't be performed in combination with other, PSyIR transformations) so I would go with it.

sergisiso commented:
@rupertford @arporter This is ready for the first review.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2721,34 +2696,28 @@ def dag_name(self):
         for kernel in my_schedule.walk(Kern):
             if kernel.name == self.name:
+                # pylint: disable=protected-access
                 kernel._module_inline = value
```
There's a setter for this so you can just do `kernel.module_inline = value`  I think?

sergisiso replied:
We are already inside this setter :) The problem is that it applies it to all kernels with the same name and if we would call the setter again it will trigger an infinite recursive call. I added a comment to make this clear.

arporter replied:
Ahem, I knew that ;-) I see the problem now. We could protect it with `if kernel is not self` and then separately set `self._module_inline = value`. I'm not sure which is best - I'll let you play and decide.

sergisiso replied:
Removed protected access.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2771,19 +2740,65 @@ def gen_code(self, parent):
         :param parent: The parent of this kernel call in the f2pygen AST.
         :type parent: :py:calls:`psyclone.f2pygen.LoopGen`
+
+        :raises NotImplementedError: if there is a name clash preventing to \
```
"...clash that prevents the kernel from being module-inlined without ..."

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2771,19 +2740,65 @@ def gen_code(self, parent):
             parent.add(UseGen(parent, name=self._module_name, only=True,
                               funcnames=[self._name]))
+        else:
+            # Find the root f2pygen module
```
Does the f2pygen ast not have a `root` method? If not, then don't worry - there's no point adding one.

sergisiso replied:
Worse than that, it does have a `root` method but it doesn't return the ancestor of the `.parent` chain.

I think it uses "root" to mean the source of the content. Anyways it returns a fparser node, while in this case I want the root f2pygen node, hence I have to do it manually.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2771,19 +2740,65 @@ def gen_code(self, parent):
+            if not existing_symbol:
+                # If it doesn't exist already, module-inline the subroutine
+                self.root.symbol_table.add(RoutineSymbol(self._name))
+                module.add(PSyIRGen(module, self.get_kernel_schedule()))
```
Please add a comment to explain that "PSyIRGen converts the Kernel PSyIR into a node in the f2pygen AST."

sergisiso replied:
Improved comments.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2771,19 +2740,65 @@ def gen_code(self, parent):
+
+                # Make sure the generated code is an exact match
+                found = False
+                search = PSyIRGen(module, self.get_kernel_schedule()).root
```
I have to admit that I don't understand this code. Can you not just do `module.root` rather than creating a `PSyIRGen` only to then look at its root? Please add comments - what does checking that the two roots are the same mean?
Also, as an implementation thing you can do away with the `found` variable:

    for child in module.children:
        if blah:
            break
    else:
         # failed to find blah

sergisiso replied:
Again I think the difficulty here could be on interpreting `root` as it doesn't return an ancestor but the fparser equivalent to the associated f2pygen node.

What I am doing is create the f2pygen  (which in turn creates the fparser1 version) of the kernel_schedule and then compare it to any of the fparser1 tree associated to f2pygen nodes under the module. If found it means there is already a subroutine with he exact same name and implementation (fparser1 tree), so we can safely continue without inserting it again.

I added a comment to try to explain all this.

Also the `for else` is a good suggestion that I always forget.

arporter replied:
Aha. I get it now. Thanks for adding comments.

arporter commented on src/psyclone/tests/domain/gocean/transformations/globalstoargs_test.py:
```diff
@@ -304,18 +306,21 @@ def create_data_symbol(arg):
 
     # Check the kernel calls have the global passed as last argument
-    assert "CALL kernel_with_use_code(i, j, oldu_fld, cu_fld%data, " \
+    assert "CALL kernel_with_use_0_code(i, j, oldu_fld, cu_fld%data, " \
```
Was this wrong before then? i.e. we didn't rename a kernel that we had applied the globals-to-args transformation to?

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -396,6 +397,115 @@ def test_codedkern_node_str():
+def test_codedkern_module_inline_gen_code(tmpdir):
+    ''' Check that a CodedKern with module-inline gets copied into the
+    local module appropriately when the PSy-layer is generated'''
+    # Use a Dynamo example that has a repeated CodedKern
```
s/Dynamo/LFRic/

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/kernel_transformation_test.py:
```diff
@@ -113,22 +113,6 @@ def test_accroutine_module_use():
             " must first" in str(err.value))
 
 
-def test_accroutine_rejects_transformed_kernel():
```
Whoop!

arporter reviewed (changes requested):
It's really nice to see so many tests removed :-)
Looks generally good but I have a question about how the check for existing subroutines of the same name works. (I'm probably being stupid because the training has pushed a lot of stuff out of my brain.)
All tests pass with compilation.
All examples OK with compilation.
Coverage is not reporting 100% but I have to confess I haven't looked to see why.
I'll check for pylint/pycodestyle next time.
Although this PR lifts the restriction associated with #490, it doesn't actually close it (since it is about re-implementing the transformation to use the PSyIR).
You probably need to update the User Guide as we (hopefully) had documented the fact that it wasn't possible to transform a kernel *and* module-inline it and now it is!

sergisiso commented on src/psyclone/f2pygen.py:
```diff
@@ -537,6 +537,25 @@ def _skip_imp_none_and_comments(self, index):
+    ''' Create a Fortran block of code from PSyIR '''
+    def __init__(self, parent, content):
+        '''
+        :param parent: node in AST to which to add the Comment as a child
```
Ops I forgot updating this. Done now.

sergisiso commented on src/psyclone/f2pygen.py:
```diff
@@ -537,6 +537,25 @@ def _skip_imp_none_and_comments(self, index):
+        from psyclone.psyir.backend.fortran import FortranWriter
+        from fparser.one.parsefortran import FortranParser
+        fortran_writer = FortranWriter()
+        reader = FortranStringReader(fortran_writer(content),
```
Added comments to this method.

sergisiso commented on src/psyclone/f2pygen.py:
```diff
@@ -537,6 +537,25 @@ def _skip_imp_none_and_comments(self, index):
+        fortran_writer = FortranWriter()
+        reader = FortranStringReader(fortran_writer(content),
+                                     ignore_comments=False)
+        fparser1_parser = FortranParser(reader, ignore_comments=False)
```
Added comments to this method.

sergisiso commented:
@arporter Ready for next review.

> Coverage is not reporting 100% but I have to confess I haven't looked to see why.

I think the coverage of the diff is 100% but since I removed lines the overall coverage has decreased, this is why codecov complains.

> Although this PR lifts the restriction associated with #490, it doesn't actually close it (since it is about re-implementing the transformation to use the PSyIR).

I see what you mean, I did mention it because I could get rid of a few 490 TODOs but to be fair there is still one remaining so I agree it should remain open.

> You probably need to update the User Guide as we (hopefully) had documented the fact that it wasn't possible to transform a kernel and module-inline it and now it is!

I updated the User Guide and also found the gocean/eg4 is testing exactly this functionality and I could unlock part of the example.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2717,6 +2717,8 @@ def module_inline(self, value):
         my_schedule = self.ancestor(InvokeSchedule)
         for kernel in my_schedule.walk(Kern):
```
Not your fault but pycodestyle reports that lines 1268-1271 are too long. Would you mind fixing?

sergisiso replied:
Fixed

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3269,6 +3246,7 @@ def apply(self, kern, options=None):
             Implicit_Part, Comment
         from fparser.two.utils import walk
         from fparser.common.readfortran import FortranStringReader
+        from psyclone.psyir.nodes import CodeBlock
```
Please move this to the top of the file.

sergisiso replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -3302,8 +3280,15 @@ def apply(self, kern, options=None):
         # Flag that the kernel has been modified
         kern.modified = True
+
+        # At the 'cmt' directive into the PSyIR as a CodeBlock
```
s/At/Add/

sergisiso replied:
Done

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -129,8 +129,6 @@ def gen(self):
         # include the field_mod module
         psy_module.add(UseGen(psy_module, name="field_mod"))
```
Not your fault but pylint complains about white space at line 71

sergisiso replied:
Fixed

arporter reviewed (changes requested):
Many thanks for the changes. I understand what's going on now! It's looking good but, just to be on the safe side, I've requested that you extend eg4 so that (given a suitable compiler) we can actually compile the OpenACC example.
Apart from that there are just a few typos.

sergisiso commented on examples/gocean/eg4/README.md:
```diff
@@ -54,27 +54,21 @@ argument:
-transformed, the other kernels that they then call would not and this
-means that they would not be compiled for execution on the OpenACC
-device. Support for recursive kernel transformation is the subject of
+However, attempting the  `alg_kern_call_kern`, although the kernels
```
Done

sergisiso commented on src/psyclone/f2pygen.py:
```diff
@@ -538,20 +539,30 @@ def _skip_imp_none_and_comments(self, index):
         fparser1_parser = FortranParser(reader, ignore_comments=False)
         fparser1_parser.parse()
+
+        # Update the f2pygen with the newly parsed fparser1 AST content
```
Done

sergisiso commented on examples/gocean/eg4/Makefile:
```diff
@@ -31,31 +31,32 @@
+	# ${PSYCLONE} -api "gocean1.0" -s ./acc_transform.py alg_kern_call_kern.f90
+	# ${PSYCLONE} -api "gocean1.0" -s ./acc_transform.py alg_kern_nested_use.f90
 
 compile:
```
Sure, it is a good idea.

I had to make some changes in the Alg and Kernel to make it at least a program and use dl_esm_inf modules properly. At the moment it just proves that it can be compiled with OpenACC, but not much else.

arporter replied:
That's great, thanks Sergi. I can build it on Glados now :-)

sergisiso commented:
@arporter Ready for next review

arporter commented on examples/gocean/eg4/README.md:
```diff
@@ -87,6 +88,21 @@ psyclone -api "gocean1.0" -s ./acc_transform.py alg_nested_use.f90
+
+Additionally, this example's Makefile has the necessary rules to compile
+the OpenAcc version of the `alg_kern_use_var.f90` version. This is done
+with the `compile-acc` target. Note that this requieres an OpenACC
```
"requires"

sergisiso replied:
Fixed

arporter commented on examples/gocean/eg4/README.md:
````diff
@@ -87,6 +88,21 @@ psyclone -api "gocean1.0" -s ./acc_transform.py alg_nested_use.f90
+be built with the same compiler.
+
+```sh
+export F90=pfg90
````
s/pfg90/pgf90/

arporter replied:
Actually, for the `compile-acc` target the Makefile assumes that we're using pgf90 so please say that the flags will need changing if using e.g. gfortran.

sergisiso replied:
Updated the text

arporter replied:
You missed the "pfg90" but I've taken the liberty of fixing that.

arporter reviewed (changes requested):
All examples compile/run OK. Ditto tests.
I can build gocean/eg4 with PGI (nvidia).
All requested changes have been made.
Just a few typos/text changes needed and then I'll merge.

sergisiso commented:
@arporter Addressed the issues and brought to master. Ready for next review
